### PR TITLE
[SPARK-29644][SQL][2.4] Corrected ShortType and ByteType mapping to SmallInt and TinyInt in JDBCUtils

### PR DESCRIPTION
**Pull request by shivsood**
This is a port SPARK-29644 to 2.4

### What changes were proposed in this pull request?

Corrected ShortType and ByteType mapping to SmallInt and TinyInt, corrected setter methods to set ShortType and ByteType as setShort() and setByte(). Changes in JDBCUtils.scala
Fixed Unit test cases to where applicable and added new E2E test cases in to test table read/write using ShortType and ByteType.

Problems
- In master in JDBCUtils.scala line number 547 and 551 have a problem where ShortType and ByteType are set as Integers rather than set as Short and Byte respectively.

```
case ShortType =>
    (stmt: PreparedStatement, row: Row, pos: Int) =>
    stmt.setInt(pos + 1, row.getShort(pos))
    The issue was pointed out by @maropu

case ByteType =>
    (stmt: PreparedStatement, row: Row, pos: Int) =>
     stmt.setInt(pos + 1, row.getByte(pos))
```

- Also at line JDBCUtils.scala 247 TinyInt is interpreted wrongly as IntergetType in getCatalystType()
``` case java.sql.Types.TINYINT => IntegerType ```

- At line 172 ShortType was wrongly interpreted as IntegerType
``` case ShortType => Option(JdbcType("INTEGER", java.sql.Types.SMALLINT)) ```

- All thru out tests, ShortType and ByteType were being interpreted as IntegerTypes.

### Why are the changes needed?
Given type should be set using the right type.

### Does this PR introduce any user-facing change?
Yes. 
- User will now be able to create tables where dataframe contains ByteType when using JDBC connector in overwrite mode.
- Users will see a SQL side table are created with the right data type. ShortType in spark will translate to smallint and ByteType to TinyInt on the SQL side. This will resulting in small size of db tables where applicable.

### How was this patch tested?
Corrected Unit test cases where applicable. Validated in CI/CD
Added/fixed test case in MsSqlServerIntegrationSuite.scala, PostgresIntegrationSuite.scala , MySQLIntegrationSuite.scala to
write/read tables from dataframe with cols as shorttype and bytetype. Validated by manual as follows.
```
./build/mvn install -DskipTests
./build/mvn test -Pdocker-integration-tests -pl :spark-docker-integration-tests_2.12
```

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by shivsood** [comment]
@dongjoon-hyun @maropu @srowen here's the PR that ports #26301 to Spark 2.4. Thanks

**Comment by dongjoon-hyun** [comment]
Yup. I've been waiting. (BTW, actually, I expected your follow-up PR on test case refactoring on master branch first.) 😄

**Comment by srowen** [comment]
I think this could be OK; it's a behavior change though? is it OK for a 2.4.x maintenance release? I guess it depends on how much we think this is more bug fix than improvement.

**Comment by dongjoon-hyun** [comment]
Yes. I agree with you, @srowen . Due to this patch, the user facing behavior will be fixed.

cc @gatorsmile and @cloud-fan

**Comment by shivsood** [comment]
> Yup. I've been waiting. (BTW, actually, I expected your follow-up PR on test case refactoring on master branch first.) smile

@dongjoon-hyun That's on my mind. Will take some time as i have to better understand the current suite and my availability next few weeks. But definitely on me to fix

**Comment by maropu** [comment]
>> Does this PR introduce any user-facing change?
> No

Can you update the description above?
Anyway, IMO this is not a serious bug fix, so it might be less worth correcting it in this branch.

**Comment by shivsood** [comment]
> Can you update the description above?
Yes,

>Anyway, IMO this is not a serious bug fix, so it might be less worth correcting it in this branch.
@maropu Originally i found the issue when validating on spark 2.4. I think its important. AFIAK without this fix, i was not able to overwrite tables in SQL server if the dataframe had a ByteType column. This was because Spark sent a JDBC create command using "BYTE" which is not a know JDBC type. The fixes in jdbcutils.scala now fix this.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by srowen** [comment]
Merged to 2.4

**Comment by dongjoon-hyun** [comment]
Thank you, @shivsood , @srowen , @maropu

**Comment by gatorsmile** [comment]
How serious is it to the end users?  I tend to agree @maropu this is not a big issue. 

You know, fixing this in the maintenance releases should be avoided. This will change the data type of the external schema. Normally, it will easily break the external applications. 

Can we revert this?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala [hunk 1]
```diff
@@ -59,7 +59,7 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationSuite {
       """
         |INSERT INTO numbers VALUES (
         |0,
-        |255, 32767, 2147483647, 9223372036854775807,
+        |127, 32767, 2147483647, 9223372036854775807,
```
Comment (HyukjinKwon) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala]: Wait, how was this possible before? Do we cover unsigned cases too?

**Comment by dongjoon-hyun** [comment]
Fixing bugs always change the behavior. All existing application which depends on the bug are affected naturally. I think this is a kind of issue of release note or migration (not a revert).

**Comment by shivsood** [comment]
> Fixing bugs always change the behavior. All existing application which depends on the bug are affected naturally. I think this is a kind of issue of release note or migration guide (not a revert).
Totally agree. Let me know where i can add documentation on corrected behavior. I have summarized in the PR description briefly.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala (hunk 1 above)
Comment (shivsood) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala]: Yes, all test cases including signed and unsigned are covered now for ByteType and ShortType. Refer to test("SPARK-29644: Write tables with ShortType") and  test("SPARK-29644: Write tables with ByteType") in JDBCWriteSuite.scala and MsSQLServerIntegrationSuite.scala.

Earlier it all worked as everything was treated as an integer. Every test cases treated ByteType and ShortType as integers. These test are corrected now.

Comment (HyukjinKwon) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala]: I meant 

1. This TINYINT seems able to contain unsigned values (https://dev.mysql.com/doc/refman/8.0/en/integer-types.html) up to 255. How do we handle?
2. Previously the value given here was `255` for `TINYINT` which is performed via MySQL if I am not mistaken. How was this possible without UNSIGNED keyword?

**Comment by srowen** [comment]
These are good questions. If this is not just a bug fix, letting things work that didn't before, it maybe shouldn't be in 2.4. That's the question at https://github.com/apache/spark/pull/26301 - we can discuss there. If there isn't a quick resolution, we can revert this pending further investigation.

**Comment by dongjoon-hyun** [comment]
Hi, All.
I'll revert this PR from `branch-2.4` due to the `ByteType` logic and the above discussion.

**Comment by dongjoon-hyun** [comment]
@shivsood . Please make a followup PR against `master` branch to revert `ByteType` change.

**Comment by srowen** [comment]
Agree, these are significant enough questions that we should reevaluate in master first. Good catch @gatorsmile ! thank you.

**Comment by shivsood** [comment]
Thanks everyone for the comments. 'll send a PR to revert the changes in master. 
cc @dongjoon-hyun @srowen @gatorsmile @HyukjinKwon @maropu

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala]: @HyukjinKwon .
It seems that we need to revert https://github.com/apache/spark/pull/23400 with the same reason. How do you think about that?

**Comment by dongjoon-hyun** [comment]
Along with this, I suggest to revert the followings which mapped SQL `TINYINT` to Spark `ByteType` for consistency.
[SPARK-26499] [SQL] JdbcUtils.makeGetter does not handle ByteType
- https://github.com/apache/spark/pull/23400 (master)
- https://github.com/apache/spark/pull/26531 (branch-2.4, I merged this.)

**Comment by HyukjinKwon** [comment]
@dongjoon-hyun, seems #23400 itself is fine although the test alone might have a potential issue. If inferred type is a bytetype, it should get a bytetype. If there's an issue in type inference in JDBC, we should fix that code path instead.

**Comment by dongjoon-hyun** [comment]
+1 for the partial revert. Only the following part is valid although `row.update` is already replaced by `row.setByte`. The test should should not be there.
```
    case ByteType =>
       (rs: ResultSet, row: InternalRow, pos: Int) =>
         row.update(pos, rs.getByte(pos + 1))
```

**Comment by dongjoon-hyun** [comment]
@HyukjinKwon and I double-checked that. Luckily, `TINYINT` of H2 database is -128 to 127. So, the `H2` dialect test code seems valid from the context. We don't need any changes.
- https://www.h2database.com/html/datatypes.html#tinyint_type

**Comment by HyukjinKwon** [comment]
Thanks for investigation.

**Pull request closed without merging**